godot-core/src/builtin/projection.rs

```diff
@@ -0,0 +1,443 @@
+        self.as_inner().get_lod_multiplier()
+    }
+
+    /// Returns the number of pixels with the given pixel width displayed per
+    /// meter, after this Projection is applied.
+    ///
+    /// _Godot equivalent: Projection.get_pixels_per_meter()_
+    pub fn pixels_per_meter(&self, pixel_width: i64) -> i64 {
+        self.as_inner().get_pixels_per_meter(pixel_width)
+    }
+
+    /// Returns the clipping plane of this Projection whose index is given by
+    /// plane.
+    ///
+    /// _Godot equivalent: Projection.get_projection_plane()_
+    pub fn projection_plane(&self, plane: ProjectionPlane) -> Plane {
+        self.as_inner().get_projection_plane(plane as i64)
+    }
+
+    /// Returns the dimensions of the viewport plane that this Projection
+    /// projects positions onto, divided by two.
+    ///
+    /// _Godot equivalent: Projection.get_viewport_half_extents()_
+    pub fn viewport_half_extents(&self) -> Vector2 {
+        self.as_inner().get_viewport_half_extents()
+    }
+
+    /// Returns the distance for this Projection beyond which positions are
+    /// clipped.
+    ///
+    /// _Godot equivalent: Projection.get_z_far()_
+    pub fn z_far(&self) -> f64 {
+        self.as_inner().get_z_far()
+    }
+
+    /// Returns the distance for this Projection before which positions are
+    /// clipped.
+    ///
+    /// _Godot equivalent: Projection.get_z_near()_
+    pub fn z_near(&self) -> f64 {
+        self.as_inner().get_z_near()
+    }
+
+    /// Returns a Projection that performs the inverse of this Projection's
+    /// projective transformation.
+    ///
+    /// _Godot equivalent: Projection.inverse()_
+    pub fn inverse(self) -> Self {
+        self.glam(|mat| mat.inverse())
+    }
+
+    /// Returns `true` if this Projection performs an orthogonal projection.
+    ///
+    /// _Godot equivalent: Projection.is_orthogonal()_
+    pub fn is_orthogonal(&self) -> bool {
+        self.as_inner().is_orthogonal()
+    }
+
+    /// Returns a Projection with the X and Y values from the given [`Vector2`]
+    /// added to the first and second values of the final column respectively.
+    ///
+    /// _Godot equivalent: Projection.jitter_offseted()_
+    pub fn jitter_offseted(&self, offset: Vector2) -> Self {
+        self.as_inner().jitter_offseted(offset)
+    }
+
+    /// Returns a Projection with the near clipping distance adjusted to be
+    /// `new_znear`.
+    ///
+    /// Note: The original Projection must be a perspective projection.
+    ///
+    /// _Godot equivalent: Projection.perspective_znear_adjusted()_
+    pub fn perspective_znear_adjusted(&self, new_znear: f64) -> Self {
+        self.as_inner().perspective_znear_adjusted(new_znear)
+    }
+
+    #[doc(hidden)]
+    pub(crate) fn as_inner(&self) -> InnerProjection {
+        InnerProjection::from_outer(self)
+    }
+}
+
+impl From<Transform3D> for Projection {
+    fn from(trans: Transform3D) -> Self {
+        trans.glam(Mat4::from)
+    }
+}
```

Comment:
Do we have a rule when we use a constructor and when a `From` conversion? 
In this case, each `Transform3D` can also act as a `Projection`, right?

Also, I assume this takes `Transform3D` by value, because `Transform3D` uses `self` for all the operations?

By the way, very elegant implementation 🙂

Reply:
To me `From` means there is a single obvious reinterpretation of the values. Such as `From<Basis>` for `Transform3D`. So yeah, i believe it does what you say. I understand some of the math behind projection matrices but im not super good at it so i can't provide a super detailed justification.

It takes it by value because `Mat4::from` does and because `Transform3D` is copy anyway. Taking `Transform3D` by reference would just lead to a deref, and therefore a copy, anyway. Internally it does seem like the `Mat4::from` function does do what you say too.